app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,367 @@
+            startTimestampInitialized = true;
+        }
+        return (System.nanoTime() / 1000 - startTimestampUs) / timeLapse;
+    }
+
+    private void startMuxerIfSetUp() {
+        if (audioTrackIndex >= 0 && videoTrackIndex >= 0) {
+            muxer.start();
+            muxerStarted = true;
+        }
+    }
+
+    @SuppressLint("NewApi")
+    @Override
+    public void run() {
+        try {
+            setupVideoCodec();
+            setupVirtualDisplay();
+
+            setupAudioCodec();
+            setupAudioRecord();
+
+            muxer = new MediaMuxer(this.outputFilePath,
+                    MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
+
+            startAudioRecord();
+
+            MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
+            long lastAudioTimestampUs = -1;
+
+            while (!stopped && !asyncError) {
+                int encoderStatus;
+
+                encoderStatus = audioEncoder.dequeueOutputBuffer(bufferInfo, 0);
+                if (encoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER) {
+                    Log.w(TAG, "audio encoder.dequeueOutputBuffer: try again");
+                } else if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+                    if (audioTrackIndex > 0) {
+                        Log.e(TAG, "audioTrackIndex less than zero");
+                        break;
+                    }
+                    audioTrackIndex = muxer.addTrack(audioEncoder.getOutputFormat());
+                    startMuxerIfSetUp();
+                } else if (encoderStatus < 0) {
+                    Log.w(TAG, "unexpected result from audio encoder.dequeueOutputBuffer: "
+                            + encoderStatus);
+                } else {
+                    ByteBuffer encodedData = audioEncoder.getOutputBuffer(encoderStatus);
+                    if (encodedData == null) {
+                        Log.e(TAG, "encodedData null");
+                        break;
+                    }
+
+                    if (bufferInfo.presentationTimeUs > lastAudioTimestampUs
+                            && muxerStarted && bufferInfo.size != 0 &&
+                            (bufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) == 0) {
+                        lastAudioTimestampUs = bufferInfo.presentationTimeUs;
+                        muxer.writeSampleData(audioTrackIndex, encodedData, bufferInfo);
+                    }
+
+                    audioEncoder.releaseOutputBuffer(encoderStatus, false);
+
+                    if ((bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
+                        Log.e(TAG, "buffer eos");
+                        break;
+                    }
+                }
+
+                if (videoTrackIndex >= 0 && audioTrackIndex < 0)
+                    continue; // wait for audio config before processing any video data frames
+
+                encoderStatus = videoEncoder.dequeueOutputBuffer(bufferInfo, 10000);
```

Comment:
what is this magic number?

Reply:
done, moved to constants, it is just for timeout value while dequeuing buffer, it blocks buffering at most "timeoutUs" microseconds so no hang